pkg/trace/sampler/exception_sampler.go

```diff
@@ -0,0 +1,205 @@
+// of env, service, name, resource, http-status, error type for each top level or measured spans
+func NewExceptionSampler() *ExceptionSampler {
+	return &ExceptionSampler{
+		limiter:   rate.NewLimiter(exceptionSamplerTPS, exceptionSamplerBurst),
+		seen:      make(map[Signature]*seenSpans),
+		tickStats: time.NewTicker(10 * time.Second),
+	}
+}
+
+// Add samples a trace and returns true if trace was sampled (should be kept)
+func (e *ExceptionSampler) Add(env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	return e.add(time.Now(), env, root, t)
+}
+
+func (e *ExceptionSampler) add(now time.Time, env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	if priority, ok := GetSamplingPriority(root); priority > 0 && ok {
+		e.handlePriorityTrace(now, env, t)
+		return false
+	}
+	return e.handleTrace(now, env, t)
+}
+
+// Start starts reporting stats
+func (e *ExceptionSampler) Start() {
+	go func() {
+		for range e.tickStats.C {
+			e.report()
+		}
+	}()
+}
+
+// Stop stops reporting stats
+func (e *ExceptionSampler) Stop() {
+	e.tickStats.Stop()
+}
+
+func (e *ExceptionSampler) handlePriorityTrace(now time.Time, env string, t pb.Trace) {
+	expire := now.Add(priorityTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+}
+
+func (e *ExceptionSampler) handleTrace(now time.Time, env string, t pb.Trace) bool {
+	var sampled bool
+	expire := now.Add(defaultTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		if !sampled {
+			sampled = e.sampleSpan(now, env, s)
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) addSpan(expire time.Time, env string, s *pb.Span) {
+	shardSig := ServiceSignature{env, s.Service}.Hash()
+	b := e.loadSpanSeenSet(shardSig)
+	b.add(expire, s)
+}
+
+func (e *ExceptionSampler) sampleSpan(now time.Time, env string, s *pb.Span) bool {
+	var sampled bool
+	shardSig := ServiceSignature{env, s.Service}.Hash()
+	b := e.loadSpanSeenSet(shardSig)
+	sig := b.sign(s)
+	expire, ok := b.getExpire(sig)
+	if now.After(expire) || !ok {
+		sampled = e.limiter.Allow()
+		if sampled {
+			b.add(now.Add(defaultTTL), s)
+			atomic.AddInt64(&e.hits, 1)
+			traceutil.SetMetric(s, exceptionKey, 1)
+		} else {
+			atomic.AddInt64(&e.misses, 1)
+		}
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) loadSpanSeenSet(shardSig Signature) *seenSpans {
+	e.mu.RLock()
+	s, ok := e.seen[shardSig]
+	e.mu.RUnlock()
+	if ok {
+		return s
+	}
+	s = &seenSpans{expires: make(map[spanHash]time.Time)}
+	e.mu.Lock()
+	e.seen[shardSig] = s
+	e.mu.Unlock()
+	return s
+}
+
+func (e *ExceptionSampler) report() {
+	metrics.Count("datadog.trace_agent.sampler.exception.hits", atomic.SwapInt64(&e.hits, 0), nil, 1)
+	metrics.Count("datadog.trace_agent.sampler.exception.misses", atomic.SwapInt64(&e.misses, 0), nil, 1)
```

Comment:
What do you think of adding a metric or a log when shrink is enabled to ease up troubleshooting?